storey/timescaledb_target.py

```diff
@@ -137,9 +165,114 @@ async def _async_init(self):
         while ensuring the pool is available when needed for data operations.
         """
         if self._pool is None:
-            import asyncpg
+            self._pool = AsyncConnectionPool(self._dsn, min_size=1, max_size=1, open=False)
+            await self._pool.open()
+
+    async def _execute_with_retry(self, operation_callable: Callable[[], Any]) -> Any:
+        """
+        Generic async retry wrapper for database operations.
+
+        PostgreSQL Error Handling Strategy Matrix:
 
-            self._pool = await asyncpg.create_pool(dsn=self._dsn, min_size=1, max_size=1)
+        | Category                    |Retry?| Timing           | Reason                           |
+        |-----------------------------|------|------------------|----------------------------------|
+        | DeadlockDetected            |  Yes | 0.1s, 0.2s, 0.4s | Auto-rollback, fast resolution  |
+        | Other OperationalError      |  Yes | 1s, 2s, 4s       | Network/server recovery time     |
+        | InterfaceError              |  Yes | 1s, 2s, 4s       | Client connection issues         |
+        | All Other psycopg.Error     |  No  | -                | Pass through without wrapping    |
+
+        Note: PostgreSQL automatically rolls back failed transactions, so explicit
+        rollback is only needed for DeadlockDetected where we retry the operation.
+
+        Note: Unhandled errors are passed through without wrapping to preserve
+        original exception types and stack traces for proper debugging.
+
+        :param operation_callable: Async function that executes the database operation
+        :return: Result of operation_callable()
+        """
+        deadlock_attempts = 0
+        connection_attempts = 0
+
+        for _ in range(self.MAX_DEADLOCK_RETRIES + self._max_retries + 1):
```

Comment:
Since it should never exit normally, it's probably clearer for it to just be
```suggestion
        while True:
```
Otherwise, it raises the question of whether this method can return `None`.

Reply:
will fix